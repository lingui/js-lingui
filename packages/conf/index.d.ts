import type { GeneratorOptions } from "@babel/core"

export declare type CatalogFormat =
  | "lingui"
  | "minimal"
  | "po"
  | "csv"
  | "po-gettext"
export type CatalogFormatOptions = {
  origins?: boolean
  lineNumbers?: boolean
}
export declare type OrderBy = "messageId" | "origin"
declare type CatalogConfig = {
  name?: string
  path: string
  include: string[]
  exclude?: string[]
}

export type LocaleObject = {
  [locale: string]: string[] | string
}

export type DefaultLocaleObject = {
  default: string
}

export declare type FallbackLocales = LocaleObject | DefaultLocaleObject

declare type CatalogService = {
  name: string
  apiKey: string
}

declare type ExtractorType = {
  match(filename: string): boolean
  extract(filename: string, targetDir: string, options?: any): void
}

export declare type LinguiConfig = {
  catalogs: CatalogConfig[]
  compileNamespace: "es" | "cjs" | "ts" | string
  extractBabelOptions: Record<string, unknown>
  compilerBabelOptions: GeneratorOptions
  fallbackLocales: FallbackLocales
  format: CatalogFormat
  extractors?: ExtractorType[]
  prevFormat: CatalogFormat
  formatOptions: CatalogFormatOptions
  localeDir: string
  locales: string[]
  catalogsMergePath: string
  orderBy: OrderBy
  pseudoLocale: string
  rootDir: string
  runtimeConfigModule:
    | [source: string, identifier?: string]
    | {
        i18n?: [source: string, identifier?: string]
        Trans?: [source: string, identifier?: string]
      }
  sourceLocale: string
  service: CatalogService
}
export declare const defaultConfig: LinguiConfig
export declare function getConfig({
  cwd,
  configPath,
  skipValidation,
}?: {
  cwd?: string
  configPath?: string
  skipValidation?: boolean
}): LinguiConfig

export declare function makeConfig(
  userConfig: Partial<LinguiConfig>,
  opts?: {
    skipValidation?: boolean
  }
): LinguiConfig

export declare const configValidation: {
  exampleConfig: {
    extractBabelOptions: {
      extends: string
      rootMode: string
      plugins: string[]
      presets: string[]
    }
    compilerBabelOptions: GeneratorOptions
    catalogs: CatalogConfig[]
    compileNamespace: "es" | "ts" | "cjs" | string
    fallbackLocales: FallbackLocales
    format: CatalogFormat
    formatOptions: CatalogFormatOptions
    locales: string[]
    orderBy: OrderBy
    pseudoLocale: string
    rootDir: string
    runtimeConfigModule: LinguiConfig["runtimeConfigModule"]
    sourceLocale: string
    service: CatalogService
  }
  deprecatedConfig: {
    fallbackLocale: (
      config: LinguiConfig & DeprecatedFallbackLanguage
    ) => string
    localeDir: (config: LinguiConfig & DeprecatedLocaleDir) => string
    srcPathDirs: (config: LinguiConfig & DeprecatedLocaleDir) => string
    srcPathIgnorePatterns: (
      config: LinguiConfig & DeprecatedLocaleDir
    ) => string
  }
  comment: string
}
export declare function replaceRootDir(
  config: LinguiConfig,
  rootDir: string
): LinguiConfig
/**
 * Replace fallbackLocale with fallbackLocales
 *
 * Released in lingui-conf 0.9
 * Remove anytime after 4.x
 */
declare type DeprecatedFallbackLanguage = {
  fallbackLocale: string | null
}

export declare function fallbackLanguageMigration(
  config: LinguiConfig & DeprecatedFallbackLanguage
): LinguiConfig
/**
 * Replace localeDir, srcPathDirs and srcPathIgnorePatterns with catalogs
 *
 * Released in @lingui/conf 3.0
 * Remove anytime after 4.x
 */
declare type DeprecatedLocaleDir = {
  localeDir: string
  srcPathDirs: string[]
  srcPathIgnorePatterns: string[]
}
export declare function catalogMigration(
  config: LinguiConfig & DeprecatedLocaleDir
): LinguiConfig
export {}
//# sourceMappingURL=index.d.ts.map
