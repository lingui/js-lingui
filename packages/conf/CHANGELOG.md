# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [5.1.2](https://github.com/lingui/js-lingui/compare/v5.1.1...v5.1.2) (2024-12-16)

**Note:** Version bump only for package @lingui/conf

## [5.1.1](https://github.com/lingui/js-lingui/compare/v5.1.0...v5.1.1) (2024-12-16)

**Note:** Version bump only for package @lingui/conf

# [5.1.0](https://github.com/lingui/js-lingui/compare/v5.0.0...v5.1.0) (2024-12-06)

**Note:** Version bump only for package @lingui/conf

## [4.14.1](https://github.com/lingui/js-lingui/compare/v4.14.0...v4.14.1) (2024-11-28)

**Note:** Version bump only for package @lingui/conf

# [4.14.0](https://github.com/lingui/js-lingui/compare/v4.13.0...v4.14.0) (2024-11-07)

**Note:** Version bump only for package @lingui/conf

# [4.13.0](https://github.com/lingui/js-lingui/compare/v4.12.0...v4.13.0) (2024-10-15)

**Note:** Version bump only for package @lingui/conf

# [4.12.0](https://github.com/lingui/js-lingui/compare/v4.11.4...v4.12.0) (2024-10-11)

### Features

- add metro transformer ([#1999](https://github.com/lingui/js-lingui/issues/1999)) ([cc7fe27](https://github.com/lingui/js-lingui/commit/cc7fe2744495e69984bf6839e217cb4216f004ce))

## [4.11.4](https://github.com/lingui/js-lingui/compare/v4.11.3...v4.11.4) (2024-09-02)

**Note:** Version bump only for package @lingui/conf

## [4.11.3](https://github.com/lingui/js-lingui/compare/v4.11.2...v4.11.3) (2024-08-09)

**Note:** Version bump only for package @lingui/conf

## [4.11.2](https://github.com/lingui/js-lingui/compare/v4.11.1...v4.11.2) (2024-07-03)

**Note:** Version bump only for package @lingui/conf

## [4.11.1](https://github.com/lingui/js-lingui/compare/v4.11.0...v4.11.1) (2024-05-30)

**Note:** Version bump only for package @lingui/conf

# [4.11.0](https://github.com/lingui/js-lingui/compare/v4.10.1...v4.11.0) (2024-05-17)

**Note:** Version bump only for package @lingui/conf

## [4.10.1](https://github.com/lingui/js-lingui/compare/v4.10.0...v4.10.1) (2024-05-03)

**Note:** Version bump only for package @lingui/conf

# [4.10.0](https://github.com/lingui/js-lingui/compare/v4.8.0...v4.10.0) (2024-04-12)

**Note:** Version bump only for package @lingui/conf

# [4.9.0](https://github.com/lingui/js-lingui/compare/v4.8.0...v4.9.0) (2024-04-12)

**Note:** Version bump only for package @lingui/conf

# [4.8.0](https://github.com/lingui/js-lingui/compare/v4.7.2...v4.8.0) (2024-04-03)

**Note:** Version bump only for package @lingui/conf

## [4.7.2](https://github.com/lingui/js-lingui/compare/v4.7.1...v4.7.2) (2024-03-26)

**Note:** Version bump only for package @lingui/conf

## [4.7.1](https://github.com/lingui/js-lingui/compare/v4.7.0...v4.7.1) (2024-02-20)

**Note:** Version bump only for package @lingui/conf

# [4.7.0](https://github.com/lingui/js-lingui/compare/v4.6.0...v4.7.0) (2024-01-05)

**Note:** Version bump only for package @lingui/conf

# [4.6.0](https://github.com/lingui/js-lingui/compare/v4.5.0...v4.6.0) (2023-12-01)

**Note:** Version bump only for package @lingui/conf

# [4.5.0](https://github.com/lingui/js-lingui/compare/v4.4.2...v4.5.0) (2023-09-14)

**Note:** Version bump only for package @lingui/conf

## [4.4.2](https://github.com/lingui/js-lingui/compare/v4.4.1...v4.4.2) (2023-08-31)

**Note:** Version bump only for package @lingui/conf

## [4.4.1](https://github.com/lingui/js-lingui/compare/v4.4.0...v4.4.1) (2023-08-30)

**Note:** Version bump only for package @lingui/conf

# [4.4.0](https://github.com/lingui/js-lingui/compare/v4.3.0...v4.4.0) (2023-08-08)

**Note:** Version bump only for package @lingui/conf

# [4.3.0](https://github.com/lingui/js-lingui/compare/v4.2.1...v4.3.0) (2023-06-29)

**Note:** Version bump only for package @lingui/conf

## [4.2.1](https://github.com/lingui/js-lingui/compare/v4.2.0...v4.2.1) (2023-06-07)

**Note:** Version bump only for package @lingui/conf

# [4.2.0](https://github.com/lingui/js-lingui/compare/v4.1.2...v4.2.0) (2023-05-26)

**Note:** Version bump only for package @lingui/conf

## [4.1.2](https://github.com/lingui/js-lingui/compare/v4.1.1...v4.1.2) (2023-05-17)

**Note:** Version bump only for package @lingui/conf

## [4.1.1](https://github.com/lingui/js-lingui/compare/v4.1.0...v4.1.1) (2023-05-17)

**Note:** Version bump only for package @lingui/conf

# [4.1.0](https://github.com/lingui/js-lingui/compare/v4.0.0...v4.1.0) (2023-05-15)

### Features

- **config:** show more clear error message when config is not found ([#1606](https://github.com/lingui/js-lingui/issues/1606)) ([d84bb8c](https://github.com/lingui/js-lingui/commit/d84bb8ce05e74e8b62be95afa0af422545b04c55))

## [3.17.2](https://github.com/lingui/js-lingui/compare/v3.17.1...v3.17.2) (2023-02-24)

**Note:** Version bump only for package @lingui/conf

## [3.17.1](https://github.com/lingui/js-lingui/compare/v3.17.0...v3.17.1) (2023-02-07)

### Bug Fixes

- **compile:** remove verbose output when using flow with template ([#1388](https://github.com/lingui/js-lingui/issues/1388)) ([31316f9](https://github.com/lingui/js-lingui/commit/31316f938957dba8e908f9f60a452a2673a934ee))
- **conf:** lazy load cosmiconfig's TypeScriptLoader ([#1403](https://github.com/lingui/js-lingui/issues/1403)) ([617a333](https://github.com/lingui/js-lingui/commit/617a3330f6eee5ffe5670a6b41f06d0a8116fc92))

# [3.17.0](https://github.com/lingui/js-lingui/compare/v3.16.1...v3.17.0) (2023-02-01)

### Bug Fixes

- **conf:** proper typescript cosmiconfig loader ([#1386](https://github.com/lingui/js-lingui/issues/1386)) ([8da122d](https://github.com/lingui/js-lingui/commit/8da122d7ee65846993a1f3b5f2091d194abef521))

## [3.16.1](https://github.com/lingui/js-lingui/compare/v3.16.0...v3.16.1) (2023-01-24)

**Note:** Version bump only for package @lingui/conf

# [3.16.0](https://github.com/lingui/js-lingui/compare/v3.15.0...v3.16.0) (2023-01-18)

**Note:** Version bump only for package @lingui/conf

# [3.15.0](https://github.com/lingui/js-lingui/compare/v3.14.0...v3.15.0) (2022-11-07)

### Features

- Switch to cosmiconfig-typescript-loader [#1247](https://github.com/lingui/js-lingui/issues/1247) ([118d183](https://github.com/lingui/js-lingui/commit/118d183805d6cfe160e6160ebd6cf6b4ed338c03))

# [3.14.0](https://github.com/lingui/js-lingui/compare/v3.13.3...v3.14.0) (2022-06-22)

**Note:** Version bump only for package @lingui/conf

## [3.13.3](https://github.com/lingui/js-lingui/compare/v3.13.2...v3.13.3) (2022-04-24)

**Note:** Version bump only for package @lingui/conf

## [3.13.2](https://github.com/lingui/js-lingui/compare/v3.13.1...v3.13.2) (2022-01-24)

**Note:** Version bump only for package @lingui/conf

## [3.13.1](https://github.com/lingui/js-lingui/compare/v3.13.0...v3.13.1) (2022-01-21)

**Note:** Version bump only for package @lingui/conf

# [3.13.0](https://github.com/lingui/js-lingui/compare/v3.12.1...v3.13.0) (2021-11-26)

**Note:** Version bump only for package @lingui/conf

## [3.12.1](https://github.com/lingui/js-lingui/compare/v3.12.0...v3.12.1) (2021-09-28)

**Note:** Version bump only for package @lingui/conf

# [3.12.0](https://github.com/lingui/js-lingui/compare/v3.11.1...v3.12.0) (2021-09-28)

### Bug Fixes

- Add missing babel properties to config validation ([#1135](https://github.com/lingui/js-lingui/issues/1135)) ([6b2d662](https://github.com/lingui/js-lingui/commit/6b2d6622b715e95b914b14df17501cc486ec45f4))

## [3.11.1](https://github.com/lingui/js-lingui/compare/v3.11.0...v3.11.1) (2021-09-07)

### Bug Fixes

- cli and conf needed hotfix for [#1110](https://github.com/lingui/js-lingui/issues/1110) ([1c6378d](https://github.com/lingui/js-lingui/commit/1c6378daa84f77757f99f6e592f6cb0e1fe02ebd))

# [3.11.0](https://github.com/lingui/js-lingui/compare/v3.10.4...v3.11.0) (2021-09-07)

### Bug Fixes

- extractors CLI validation and accept require and require.resolve ([#1126](https://github.com/lingui/js-lingui/issues/1126)) ([c50a06e](https://github.com/lingui/js-lingui/commit/c50a06eac65d71853ec2c29905a0c4883a5cc70d))
- fix testing suite [#1098](https://github.com/lingui/js-lingui/issues/1098) ([1144cc7](https://github.com/lingui/js-lingui/commit/1144cc782b25113366245922131e58bc40b2047d))
- sideEffects to false for shrinking bundle size ([#1129](https://github.com/lingui/js-lingui/issues/1129)) ([57cd2e5](https://github.com/lingui/js-lingui/commit/57cd2e576945ba30aea30d5cf5bcb27d1f77fe4c))
- unicode chars were removed from .po files when extracting ([#1125](https://github.com/lingui/js-lingui/issues/1125)) ([d688329](https://github.com/lingui/js-lingui/commit/d688329c0457a080d5cd485a95e94ffa5e00c5ae))

### Features

- Cloud service providers & Translation.io ([#1107](https://github.com/lingui/js-lingui/issues/1107)) ([cbc87b5](https://github.com/lingui/js-lingui/commit/cbc87b5dab8daf0cd2217b2a2525dfd12dad7272))

## [3.10.4](https://github.com/lingui/js-lingui/compare/v3.10.3...v3.10.4) (2021-06-16)

**Note:** Version bump only for package @lingui/conf

## [3.10.3](https://github.com/lingui/js-lingui/compare/v3.10.2...v3.10.3) (2021-06-14)

**Note:** Version bump only for package @lingui/conf

## [3.10.2](https://github.com/lingui/js-lingui/compare/v3.10.1...v3.10.2) (2021-06-08)

**Note:** Version bump only for package @lingui/conf

## [3.10.1](https://github.com/lingui/js-lingui/compare/v3.10.0...v3.10.1) (2021-06-08)

**Note:** Version bump only for package @lingui/conf

# [3.10.0](https://github.com/lingui/js-lingui/compare/v3.9.0...v3.10.0) (2021-06-08)

**Note:** Version bump only for package @lingui/conf

# [3.9.0](https://github.com/lingui/js-lingui/compare/v3.8.10...v3.9.0) (2021-05-18)

### Features

- configurable and extendable extractors with Lingui config ([#1065](https://github.com/lingui/js-lingui/issues/1065)) ([263ee59](https://github.com/lingui/js-lingui/commit/263ee59163c94220f5ba3999aa60ca40dc352e0c))

## [3.8.10](https://github.com/lingui/js-lingui/compare/v3.8.9...v3.8.10) (2021-04-19)

**Note:** Version bump only for package @lingui/conf

## [3.8.9](https://github.com/lingui/js-lingui/compare/v3.8.8...v3.8.9) (2021-04-09)

**Note:** Version bump only for package @lingui/conf

## [3.8.8](https://github.com/lingui/js-lingui/compare/v3.8.7...v3.8.8) (2021-04-09)

**Note:** Version bump only for package @lingui/conf

## [3.8.7](https://github.com/lingui/js-lingui/compare/v3.8.6...v3.8.7) (2021-04-09)

**Note:** Version bump only for package @lingui/conf

## [3.8.6](https://github.com/lingui/js-lingui/compare/v3.8.5...v3.8.6) (2021-04-08)

**Note:** Version bump only for package @lingui/conf

## [3.8.5](https://github.com/lingui/js-lingui/compare/v3.8.4...v3.8.5) (2021-04-08)

**Note:** Version bump only for package @lingui/conf

## [3.8.4](https://github.com/lingui/js-lingui/compare/v3.8.3...v3.8.4) (2021-04-08)

### Bug Fixes

- typescript loader config ([#1029](https://github.com/lingui/js-lingui/issues/1029)) ([93afb72](https://github.com/lingui/js-lingui/commit/93afb72c1d6235df1516655d5ab4eea0eeb606ee))

## [3.8.3](https://github.com/lingui/js-lingui/compare/v3.8.2...v3.8.3) (2021-04-05)

**Note:** Version bump only for package @lingui/conf

## [3.8.2](https://github.com/lingui/js-lingui/compare/v3.8.1...v3.8.2) (2021-03-31)

**Note:** Version bump only for package @lingui/conf

## [3.8.1](https://github.com/lingui/js-lingui/compare/v3.8.0...v3.8.1) (2021-03-23)

### Bug Fixes

- fallback locales default not overwritten ([78e4576](https://github.com/lingui/js-lingui/commit/78e45766fda0c147b77f96d71e9d775743ee8d18))

# [3.8.0](https://github.com/lingui/js-lingui/compare/v3.7.2...v3.8.0) (2021-03-23)

### Features

- allow to disable lineNumbers ([#1007](https://github.com/lingui/js-lingui/issues/1007)) ([fe67e0f](https://github.com/lingui/js-lingui/commit/fe67e0f7986188bff2c102703c4df3507506e0f2))

## [3.7.2](https://github.com/lingui/js-lingui/compare/v3.7.1...v3.7.2) (2021-03-14)

**Note:** Version bump only for package @lingui/conf

## [3.7.1](https://github.com/lingui/js-lingui/compare/v3.7.0...v3.7.1) (2021-03-07)

**Note:** Version bump only for package @lingui/conf

# [3.7.0](https://github.com/lingui/js-lingui/compare/v3.6.0...v3.7.0) (2021-03-04)

**Note:** Version bump only for package @lingui/conf

# [3.6.0](https://github.com/lingui/js-lingui/compare/v3.5.1...v3.6.0) (2021-02-23)

**Note:** Version bump only for package @lingui/conf

## [3.5.1](https://github.com/lingui/js-lingui/compare/v3.5.0...v3.5.1) (2021-02-09)

### Bug Fixes

- runtimeConfigModule config validation ([#972](https://github.com/lingui/js-lingui/issues/972)) ([5656c95](https://github.com/lingui/js-lingui/commit/5656c95b8b2ecfdfe903a941918c60abe16e1691))

# [3.5.0](https://github.com/lingui/js-lingui/compare/v3.4.0...v3.5.0) (2021-02-02)

### Bug Fixes

- @lingui/conf lodash.get dependency ([#950](https://github.com/lingui/js-lingui/issues/950)) ([f7b59ab](https://github.com/lingui/js-lingui/commit/f7b59abbcfc8bd797478a549641ca9e3b97a9701))

# [3.4.0](https://github.com/lingui/js-lingui/compare/v3.3.0...v3.4.0) (2021-01-13)

### Bug Fixes

- compilerBabelOptions warning with jest-validate ([#938](https://github.com/lingui/js-lingui/issues/938)) ([087ec1f](https://github.com/lingui/js-lingui/commit/087ec1fe45076bcfc8ea3a6f7657f39bbe1c1d58))

### Features

- Lingui compiles to typescript files ([#942](https://github.com/lingui/js-lingui/issues/942)) ([10bce7d](https://github.com/lingui/js-lingui/commit/10bce7dc890f208dc71bcf81dc34e57d389544fe))
- Support lingui config in multiple formats ([#941](https://github.com/lingui/js-lingui/issues/941)) ([8631111](https://github.com/lingui/js-lingui/commit/8631111a83c5f2fb87e7b6d794a279dbf7e8579d))

# [3.3.0](https://github.com/lingui/js-lingui/compare/v3.2.3...v3.3.0) (2020-12-08)

### Features

- add support for runtimeConfigModule w/ Trans ([#895](https://github.com/lingui/js-lingui/issues/895)) ([23b06b5](https://github.com/lingui/js-lingui/commit/23b06b5dfbf0db306cdfab83801898caceb5a8b0))
- config accepts compilerBabelOptions ([#906](https://github.com/lingui/js-lingui/issues/906)) ([38d01ef](https://github.com/lingui/js-lingui/commit/38d01ef13a7867460b68ab709f94a17176a21f25))
- Implement gettext plurals for PO files ([#677](https://github.com/lingui/js-lingui/issues/677)) ([415b90e](https://github.com/lingui/js-lingui/commit/415b90e0abfb24bb7170a5ba7630a4ead94898dd))

## [3.2.3](https://github.com/lingui/js-lingui/compare/v3.2.2...v3.2.3) (2020-11-22)

**Note:** Version bump only for package @lingui/conf

## [3.2.2](https://github.com/lingui/js-lingui/compare/v3.2.1...v3.2.2) (2020-11-20)

### Bug Fixes

- fallbackLocales overriden if parent found ([a53e12f](https://github.com/lingui/js-lingui/commit/a53e12f19cebeb6412debc9dace0b4a45aa17624))

# [3.2.0](https://github.com/lingui/js-lingui/compare/v3.1.0...v3.2.0) (2020-11-12)

**Note:** Version bump only for package @lingui/conf

# [3.1.0](https://github.com/lingui/js-lingui/compare/v3.0.3...v3.1.0) (2020-11-10)

### Features

- use fallback locales from cldr ([#820](https://github.com/lingui/js-lingui/issues/820)) ([2d9e124](https://github.com/lingui/js-lingui/commit/2d9e124b91f1ba7a65e9f997a3ba952679c6c23a))
