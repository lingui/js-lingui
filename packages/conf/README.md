[![License][badge-license]][license]
[![Version][badge-version]][package]
[![Downloads][badge-downloads]][package]

# @lingui/conf

> Get lingui configuration from package.json

**⚠️ Internal package: You probably don't need this.**

`@lingui/conf` is part of [LinguiJS][linguijs]. See the [documentation][documentation] for all information, tutorials and examples.

Package finds nearest package.json starting from current directory, reads `lingui` configuration, provides defaults for all options and `<rootDir>` with current working directory.

## Usage

```js
const getConfig = require("@lingui/conf")

const config = getConfig()
// When package.json is located in '/app/package.json'
// config.rootDir === '/app'
// config.localeDir === '/app/locale'
```

See the [reference][reference] documentation of `lingui-cli` for all options.

## License

This package is licensed under [MIT][license] license.

[license]: https://github.com/lingui/js-lingui/blob/main/LICENSE
[linguijs]: https://github.com/lingui/js-lingui
[documentation]: https://lingui.dev
[reference]: https://lingui.dev/ref/conf
[package]: https://www.npmjs.com/package/@lingui/conf
[badge-downloads]: https://img.shields.io/npm/dw/@lingui/conf.svg
[badge-version]: https://img.shields.io/npm/v/@lingui/conf.svg
[badge-license]: https://img.shields.io/npm/l/@lingui/conf.svg
