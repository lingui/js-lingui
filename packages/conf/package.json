{
  "name": "@lingui/conf",
  "version": "0.0.0-managed-by-release-script",
  "description": "Get lingui configuration from package.json",
  "keywords": [
    "lingui",
    "config"
  ],
  "repository": "lingui/js-lingui",
  "bugs": "https://github.com/lingui/js-lingui/issues",
  "license": "MIT",
  "author": {
    "name": "Tomáš Ehrlich",
    "email": "tomas.ehrlich@gmail.com"
  },
  "main": "index.js",
  "engines": {
    "node": ">=4.0"
  },
  "dependencies": {
    "chalk": "^2.3.0",
    "cosmiconfig": "^4.0.0",
    "jest-regex-util": "^23.3.0",
    "jest-validate": "^23.3.0",
    "pkg-conf": "^2.1.0"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.js"
  ]
}
