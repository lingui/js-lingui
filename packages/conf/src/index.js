const path = require("path")
const chalk = require("chalk")
const cosmiconfig = require("cosmiconfig")
const { validate } = require("jest-validate")
const { replacePathSepForRegex } = require("jest-regex-util")

const NODE_MODULES = replacePathSepForRegex(
  path.sep + "node_modules" + path.sep
)

export function replaceRootDir(conf, rootDir) {
  const replace = s => s.replace("<rootDir>", rootDir)
  ;["srcPathDirs", "srcPathIgnorePatterns", "localeDir"].forEach(key => {
    const value = conf[key]

    if (!value) {
    } else if (typeof value === "string") {
      conf[key] = replace(value)
    } else if (value.length) {
      conf[key] = value.map(replace)
    }
  })

  conf.rootDir = rootDir
  return conf
}

export const defaultConfig = {
  localeDir: "./locale",
  sourceLocale: "",
  fallbackLocale: "",
  srcPathDirs: ["<rootDir>"],
  srcPathIgnorePatterns: [NODE_MODULES],
  format: "lingui",
  rootDir: ".",
  extractBabelOptions: {
	  plugins: [],
	  presets: []
  }
}

const deprecatedConfig = {
  fallbackLanguage: config =>
    ` Option ${chalk.bold("fallbackLanguage")} was replaced by ${chalk.bold(
      "fallbackLocale"
    )}
    
    @lingui/cli now treats your current configuration as:
    {
      ${chalk.bold('"fallbackLocale"')}: ${chalk.bold(
      `"${config.fallbackLanguage}"`
    )}
    }
    
    Please update your configuration.
    `
}

const configValidation = {
  exampleConfig: defaultConfig,
  deprecatedConfig,
  comment: "See https://lingui.js.org/ref/conf.html for a list of valid options"
}

export function getConfig({ cwd } = {}) {
  const defaultRootDir = cwd || process.cwd()
  const configExplorer = cosmiconfig("lingui")
  const result = configExplorer.searchSync(defaultRootDir)
  const raw = Object.assign({}, defaultConfig, result ? result.config : null)

  validate(raw, configValidation)
  // Use deprecated fallbackLanguage, if defined
  raw.fallbackLocale = raw.fallbackLocale || raw.fallbackLanguage || ""

  const rootDir = result ? path.dirname(result.filepath) : defaultRootDir
  return replaceRootDir(raw, rootDir)
}
