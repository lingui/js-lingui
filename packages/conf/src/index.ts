export { makeConfig } from "./makeConfig"
export { getConfig } from "./getConfig"
export * from "./types"
