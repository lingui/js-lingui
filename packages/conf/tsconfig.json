{
  "compilerOptions": {
    "module": "commonjs",
    "target": "ES2019",
    "sourceMap": true,
    "noEmit": false,
    "declaration": true,
    "outDir": "./build",
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "strictNullChecks": false
  },
  "files": [
    "./src/index.ts"
  ],
  "exclude": [
    "build",
    "node_modules"
  ]
}
