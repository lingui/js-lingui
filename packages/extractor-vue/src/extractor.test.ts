import { makeConfig } from "@lingui/conf"
import fs from "fs"
import path from "path"
import { vueExtractor } from "."
import type { ExtractedMessage } from "@lingui/babel-plugin-extract-messages"

function normalizePath(entries: ExtractedMessage[]): ExtractedMessage[] {
  return entries.map((entry) => {
    const [filename, lineNumber, column] = entry.origin
    const projectRoot = process.cwd()

    return {
      ...entry,
      origin: [path.relative(projectRoot, filename ?? ""), lineNumber, column],
    }
  })
}

describe("vue extractor", () => {
  const linguiConfig = makeConfig({
    locales: ["en", "nb"],
    sourceLocale: "en",
    rootDir: ".",
    catalogs: [
      {
        path: "<rootDir>/{locale}",
        include: ["<rootDir>"],
        exclude: [],
      },
    ],
    extractorParserOptions: {
      tsExperimentalDecorators: false,
      flow: false,
    },
  })

  it("should ignore non vue files in extractor", async () => {
    const match = vueExtractor.match("test.js")

    expect(match).toBeFalsy()
  })

  it("should extract message from vue file", async () => {
    const filePath = path.resolve(__dirname, "fixtures/test.vue")
    const code = fs.readFileSync(filePath, "utf-8")

    let messages: ExtractedMessage[] = []

    await vueExtractor.extract(
      "test.vue",
      code,
      (res) => {
        messages.push(res)
      },
      {
        linguiConfig,
      }
    )

    messages = normalizePath(messages)

    expect(messages).toMatchSnapshot()
  })

  it("should extract message from functional component", async () => {
    const filePath = path.resolve(__dirname, "fixtures/functional.vue")
    const code = fs.readFileSync(filePath, "utf-8")

    let messages: ExtractedMessage[] = []

    await vueExtractor.extract(
      "functional.vue",
      code,
      (res) => {
        messages.push(res)
      },
      {
        linguiConfig,
      }
    )

    messages = normalizePath(messages)

    expect(messages).toMatchSnapshot()
  })
})
