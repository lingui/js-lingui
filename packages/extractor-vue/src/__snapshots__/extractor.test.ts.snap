// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vue extractor should extract message from functional component 1`] = `
[
  {
    comment: undefined,
    context: undefined,
    id: Render function message,
    message: undefined,
    origin: [
      functional.vue,
      10,
      33,
    ],
  },
]
`;

exports[`vue extractor should extract message from vue file 1`] = `
[
  {
    comment: undefined,
    context: undefined,
    id: Setup message,
    message: undefined,
    origin: [
      test.vue.ts,
      4,
      0,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: Script message,
    message: undefined,
    origin: [
      test.vue,
      19,
      20,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: custom.id,
    message: My message,
    origin: [
      test.vue,
      27,
      11,
    ],
  },
  {
    comment: Message comment,
    context: undefined,
    id: my.message,
    message: My descriptor message,
    origin: [
      test.vue,
      29,
      10,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: id used as message,
    message: undefined,
    origin: [
      test.vue,
      35,
      5,
    ],
  },
  {
    comment: undefined,
    context: undefined,
    id: My message without ID and context,
    message: undefined,
    origin: [
      test.vue,
      36,
      11,
    ],
  },
]
`;
